kernel/xpmem_attach.c

```diff
@@ -162,36 +162,109 @@ xpmem_close_handler(struct vm_area_struct *vma)
 #endif
 }
 
-static void
+/* Find the range of non present entries */
+static u64
+xpmem_fault_vaddr_end(u64 vm_end, u64 addr, unsigned long max_pages)
+{
+	for (; max_pages; max_pages--, addr += PAGE_SIZE) {
+		if (addr >= vm_end) {
+			break;
+		}
+		if (xpmem_vaddr_to_PFN(current->mm, addr)) {
+			break;
+		}
+	}
+	return addr;
+}
+
+/* Find the range of non present entries before the request address */
+static u64
+xpmem_fault_vaddr_start(u64 vm_start, u64 addr, unsigned long max_pages)
+{
+	for (; max_pages; max_pages--, addr -= PAGE_SIZE) {
+		if (addr <= vm_start) {
+			break;
+		}
+		if (xpmem_vaddr_to_PFN(current->mm, addr - PAGE_SIZE)) {
```

Comment:
why need to check `addr - PAGE_SIZE`?

Reply:
The `addr` is the first virtual address included in the range. The check means that if the previous virtual address has a page mapped (returns non-null pfn), we break and don't include it in the range.

In case of forward faulting scheme, this avoids repeatedly trying to fault in previous pages.

Reply:
ic